automates/model_analysis/graph_manipulation.py

```diff
@@ -0,0 +1,538 @@
+    stack_top = len(x)
+    stack_size = max(stack_top, 64)
+    stack = [False] * stack_size
+    stack[0:stack_top] = [True] * len(range(0, stack_top))
+    stack_names = [None] * stack_size
+    stack_names[0:stack_top] = copy.deepcopy(x)
+    visited_top = 0
+    visited_size = 64
+    visited = [False] * visited_size
+    visited_names = [None] * visited_size
+    is_visited = False
+    while stack_top > 0:
+        is_visited = False
+        el = stack[stack_top - 1]
+        el_name = stack_names[stack_top - 1]
+        stack_top = stack_top - 1
+        if visited_top > 0:
+            for i in range(0, visited_top):
+                if el == visited[i] and el_name == visited_names[i]:
+                    is_visited = True
+                    break
+        if not is_visited:
+            if el_name in y:
+                return False
+            visited_top = visited_top + 1
+            if visited_top > visited_size:
+                visited_old = copy.deepcopy(visited)
+                visited_size_old = visited_size
+                visited_names_old = copy.deepcopy(visited_names)
+                visited_size = 2 * visited_size
+                visited = [False] * visited_size
+                visited[0:visited_size_old] = copy.deepcopy(visited_old)
+                visited_names = [None] * visited_size
+                visited_names[0:visited_size_old] = copy.deepcopy(visited_names_old)
+            visited[visited_top - 1] = el
+            visited_names[visited_top - 1] = el_name
+            if el and (not (el_name in z)):
+                visitable_parents = list((set(parents_unsort([el_name], g)) - set([el_name])) & set(an_xyz))
+                visitable_children = list((set(children_unsort([el_name], g)) - set([el_name])) & set(an_xyz))
+                n_vis_pa = len(visitable_parents)
+                n_vis_ch = len(visitable_children)
+                if n_vis_pa + n_vis_ch > 0:
+                    while n_vis_pa + n_vis_ch + stack_top > stack_size:
+                        stack_old = copy.deepcopy(stack)
+                        stack_names_old = copy.deepcopy(stack_names)
+                        stack_size_old = stack_size
+                        stack_size = 2 * stack_size
+                        stack = [False] * stack_size
+                        stack[0:stack_size_old] = copy.deepcopy(stack_old)
+                        stack_names = [None] * stack_size
+                        stack_names[0:stack_size_old] = copy.deepcopy(stack_names_old)
+                    stack_add = stack_top + n_vis_pa + n_vis_ch
+                    stack[stack_top:stack_add] = [True] * n_vis_pa + [False] * n_vis_ch
+                    stack_names[stack_top:stack_add] = copy.deepcopy(visitable_parents) + copy.deepcopy(
+                        visitable_children)
+                    stack_top = stack_add
+            elif not el:
+                if not (el_name in z):
```

Comment:
@aswinchester the operation `(el_name in z)` is a linear time operation. That's an expensive op for a conditional evaluation and you have now done it twice (in the cases where we reach the `elif`). Can you assign the result of that expression to a variable and then use that variable in both conditional evaluation?

Reply:
This has been completed